open-api/lang-tools/src/main/java/io/deephaven/lang/completion/CompletionRequest.java

```diff
@@ -79,11 +85,101 @@ public TableDefinition getTableDefinition(VariableProvider variables, String nam
             // This might seem a little excessive, but in python at least, it is non-"free" to check if binding a variable exists.
             return localDefs.get(name);
         }
-        final TableDefinition result = variables.getTableDefinition(name);
+        TableDefinition result = variables.getTableDefinition(name);
+        if (result == null) {
+            // If the result was null, we can try to search for an assign statement that is initialized w/ something we _can_ grok.
+            final List<ChunkerAssign> assignment = completer.findAssignment(doc, this, name);
+            if (!assignment.isEmpty()) {
+                // ok! there was an assignment to our table variable that occurred before user's cursor (this is,
+                // of course, bad when user creates some random functions that are defined in any order).
+                final ListIterator<ChunkerAssign> itr = assignment.listIterator(assignment.size());
+                while (itr.hasPrevious()) {
+                    final ChunkerAssign check = itr.previous();
+                    result = findTableDefFromAssign(check);
+                    if (result != null) {
+                        break;
+                    }
+                }
+            }
+        }
         localDefs.put(name, result);
         return result;
     }
 
+    private TableDefinition findTableDefFromAssign(final ChunkerAssign check) {
+        final Node value = check.getValue();
+        if (value instanceof ChunkerInvoke) {
+            ChunkerInvoke invoke = (ChunkerInvoke) value;
+            if ("newTable".equals(invoke.getName())) {
+                // TODO: consider checking the scope of newTable... for now, too messy for ~0 value.
+                return convertNewTableInvocation(invoke);
+            }
+        }
+        return null;
+    }
+
+    private TableDefinition convertNewTableInvocation(final ChunkerInvoke invoke) {
+        final TableDefinition def = new TableDefinition();
+        final List<ColumnDefinition<?>> columns = new ArrayList<>();
+        for (Node argument : invoke.getArguments()) {
+            if (argument instanceof ChunkerInvoke) {
+                final ChunkerInvoke colInvoke = ((ChunkerInvoke) argument);
+                String colMethod = colInvoke.getName();
+                final List<Node> colArgs = colInvoke.getArguments();
+                final String colName = colArgs.isEmpty() ? null : toStringLiteral(colArgs.get(0));
```

Comment:
I think `colName == null` shouldn't produce a valid `ColumnDefinition`...

Reply:
Well... the thing about autocomplete is that it's expected to be functional even if the code is not valid.
example:
`myTable = newTable(stringCol())`
There's no arguments yet.  We know it is a string column, but that's it. 

Perhaps we _should_ elide it, since suggestions later on for myTable would only know the column type and without a name, that's not really useful.

Do you think I should just skip when there's no args?

Reply:
I mean that if you try to produce a column definition w/ colName == null, you'll get an NPE from ColumnDefinition constructor.

Reply:
ok, I am no longer allowing nulls / empty through.